File: cc/jwt/internal/jwt_mac_wrapper.cc
// Copyright 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
///////////////////////////////////////////////////////////////////////////////

#include "tink/jwt/internal/jwt_mac_wrapper.h"

#include <string>
#include <utility>

#include "absl/status/status.h"
#include "tink/jwt/internal/jwt_format.h"
#include "tink/jwt/internal/jwt_mac_internal.h"
#include "tink/jwt/jwt_mac.h"
#include "tink/primitive_set.h"
#include "tink/util/status.h"
#include "tink/util/statusor.h"

namespace crypto {
namespace tink {
namespace jwt_internal {

using ::google::crypto::tink::OutputPrefixType;

namespace {

class JwtMacSetWrapper : public JwtMac {
 public:
  explicit JwtMacSetWrapper(
      std::unique_ptr<PrimitiveSet<JwtMacInternal>> jwt_mac_set)
      : jwt_mac_set_(std::move(jwt_mac_set)) {}

  crypto::tink::util::StatusOr<std::string> ComputeMacAndEncode(
      const crypto::tink::RawJwt& token) const override;

  crypto::tink::util::StatusOr<crypto::tink::VerifiedJwt> VerifyMacAndDecode(
      absl::string_view compact,
      const crypto::tink::JwtValidator& validator) const override;

  ~JwtMacSetWrapper() override {}

 private:
  std::unique_ptr<PrimitiveSet<JwtMacInternal>> jwt_mac_set_;
};

util::Status Validate(PrimitiveSet<JwtMacInternal>* jwt_mac_set) {
  if (jwt_mac_set == nullptr) {
    return util::Status(absl::StatusCode::kInternal,
                        "jwt_mac_set must be non-NULL");
  }
  if (jwt_mac_set->get_primary() == nullptr) {
    return util::Status(util::error::INVALID_ARGUMENT,
                        "jwt_mac_set has no primary");
  }
  for (const auto* entry : jwt_mac_set->get_all()) {
    if ((entry->get_output_prefix_type() != OutputPrefixType::RAW) &&
        (entry->get_output_prefix_type() != OutputPrefixType::TINK)) {
      return util::Status(util::error::INVALID_ARGUMENT,
                          "all JWT keys must be either RAW or TINK");
    }
  }
  return util::OkStatus();
}

util::StatusOr<std::string> JwtMacSetWrapper::ComputeMacAndEncode(
    const crypto::tink::RawJwt& token) const {
  auto primary = jwt_mac_set_->get_primary();
  absl::optional<std::string> kid =
      GetKid(primary->get_key_id(), primary->get_output_prefix_type());
  return primary->get_primitive().ComputeMacAndEncodeWithKid(token, kid);
}

util::StatusOr<crypto::tink::VerifiedJwt> JwtMacSetWrapper::VerifyMacAndDecode(
    absl::string_view compact,
    const crypto::tink::JwtValidator& validator) const {
  absl::optional<util::Status> interesting_status;
  for (const auto* mac_entry : jwt_mac_set_->get_all()) {
    JwtMacInternal& jwt_mac = mac_entry->get_primitive();
    absl::optional<std::string> kid =
        GetKid(mac_entry->get_key_id(), mac_entry->get_output_prefix_type());
    util::StatusOr<VerifiedJwt> verified_jwt =
        jwt_mac.VerifyMacAndDecodeWithKid(compact, validator, kid);
    if (verified_jwt.ok()) {
      return verified_jwt;
    } else if (verified_jwt.status().code() !=
               absl::StatusCode::kUnauthenticated) {
      // errors that are not the result of a MAC verification
      interesting_status = verified_jwt.status();
    }
  }
  if (interesting_status.has_value()) {
    return *interesting_status;
  }
  return util::Status(util::error::INVALID_ARGUMENT, "verification failed");
}

}  // namespace

util::StatusOr<std::unique_ptr<JwtMac>> JwtMacWrapper::Wrap(
    std::unique_ptr<PrimitiveSet<JwtMacInternal>> jwt_mac_set) const {
  util::Status status = Validate(jwt_mac_set.get());
  if (!status.ok()) return status;
  std::unique_ptr<JwtMac> jwt_mac =
      absl::make_unique<JwtMacSetWrapper>(std::move(jwt_mac_set));
  return std::move(jwt_mac);
}

}  // namespace jwt_internal
}  // namespace tink
}  // namespace crypto

File: java_src/src/main/java/com/google/crypto/tink/jwt/JwtMacWrapper.java
// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
////////////////////////////////////////////////////////////////////////////////

package com.google.crypto.tink.jwt;

import com.google.crypto.tink.PrimitiveSet;
import com.google.crypto.tink.PrimitiveWrapper;
import com.google.crypto.tink.Registry;
import com.google.crypto.tink.proto.OutputPrefixType;
import com.google.errorprone.annotations.Immutable;
import java.security.GeneralSecurityException;
import java.util.List;
import java.util.Optional;

/**
 * JwtMacWrapper is the implementation of {@link PrimitiveWrapper} for the {@link JwtMac} primitive.
 */
class JwtMacWrapper implements PrimitiveWrapper<JwtMacInternal, JwtMac> {
  private static void validate(PrimitiveSet<JwtMacInternal> primitiveSet)
      throws GeneralSecurityException {
    if (primitiveSet.getPrimary() == null) {
      throw new GeneralSecurityException("Primitive set has no primary.");
    }
    for (List<PrimitiveSet.Entry<JwtMacInternal>> entries : primitiveSet.getAll()) {
      for (PrimitiveSet.Entry<JwtMacInternal> entry : entries) {
        if ((entry.getOutputPrefixType() != OutputPrefixType.RAW)
            && (entry.getOutputPrefixType() != OutputPrefixType.TINK)) {
          throw new GeneralSecurityException("unsupported OutputPrefixType");
        }
      }
    }
  }

  @Immutable
  private static class WrappedJwtMac implements JwtMac {
    @SuppressWarnings("Immutable") // We do not mutate the primitive set.
    private final PrimitiveSet<JwtMacInternal> primitives;

    private WrappedJwtMac(PrimitiveSet<JwtMacInternal> primitives) {
      this.primitives = primitives;
    }

    @Override
    public String computeMacAndEncode(RawJwt token) throws GeneralSecurityException {
      PrimitiveSet.Entry<JwtMacInternal> entry = primitives.getPrimary();
      Optional<String> kid = JwtFormat.getKid(entry.getKeyId(), entry.getOutputPrefixType());
      return entry.getPrimitive().computeMacAndEncodeWithKid(token, kid);
    }

    @Override
    public VerifiedJwt verifyMacAndDecode(String compact, JwtValidator validator)
        throws GeneralSecurityException {
      GeneralSecurityException interestingException = null;
      for (List<PrimitiveSet.Entry<JwtMacInternal>> entries : primitives.getAll()) {
        for (PrimitiveSet.Entry<JwtMacInternal> entry : entries) {
          try {
            Optional<String> kid = JwtFormat.getKid(entry.getKeyId(), entry.getOutputPrefixType());
            return entry.getPrimitive().verifyMacAndDecodeWithKid(compact, validator, kid);
          } catch (GeneralSecurityException e) {
            if (e instanceof JwtInvalidException) {
              // Keep this exception so that we are able to throw a meaningful message in the end
              interestingException = e;
            }
            // Ignored as we want to continue verification with other raw keys.
          }
        }
      }
      if (interestingException != null) {
        throw interestingException;
      }
      throw new GeneralSecurityException("invalid MAC");
    }
  }

  JwtMacWrapper() {}

  @Override
  public JwtMac wrap(final PrimitiveSet<JwtMacInternal> primitives)
      throws GeneralSecurityException {
    validate(primitives);
    return new WrappedJwtMac(primitives);
  }

  @Override
  public Class<JwtMac> getPrimitiveClass() {
    return JwtMac.class;
  }

  @Override
  public Class<JwtMacInternal> getInputPrimitiveClass() {
    return JwtMacInternal.class;
  }

 public static void register() throws GeneralSecurityException {
    Registry.registerPrimitiveWrapper(new JwtMacWrapper());
  }
}
